src/python/pants/backend/python/util_rules/pex_requirements.py

```diff
@@ -85,13 +87,17 @@ class LoadedLockfile:
     # The original file or file content (which may not have identical content to the output
     # `lockfile_digest`).
     original_lockfile: Lockfile | LockfileContent
+    # The parsed contents of the lockfile, if requested else None.
+    # N.B. the structure of this data is ~opaque to Pants, so use with care.
```

Comment:
Also, I've realized that it's a little weird for LoadedLockfile to have this field, when this data is already available in its lockfile_digest field. Why not parse out of that digest at the site of use. i.e., inside `_generate_python_lockfile_diff`, rather than passing around a field that isn't needed by any other consumer of this class?

Reply:
> Do we mean to say that code using this data must be aware that the format is not controlled by Pants and may change?

Yes. That. @jsirois keeps reminding us that the pex lockfile format is an undocumented implementation detail that may change.

> Why not parse out of that digest at the site of use.

It's two-fold. One is to avoid duplicating the logic to separate the Pants metadata header from the lockfile contents. This could be refactored into some helper/util function. Two is I see this information will be useful in other scenarios in the future so we want one source for parsing this rather than duplicating it for each use case. For instance this: https://github.com/pantsbuild/pants/issues/12733#issuecomment-1322192838

Reply:
In that case, "opaque" is the wrong term to use here. Opaque would mean "we don't look inside it". But in fact we do. So what we want to say is that the format may change out from under us.

Reply:
And re parsing the digest at the site of use: If I understand correctly, the lockfile_digest already contains the stripped lockfile, so it is already exactly the data you want to json-parse, no? It seems weird to pass around a json representation of a digest next to the digest, when the receiver can do json parsing as needed.

And if in the future we need this data and don't want to parse it twice, well, that's what rule memoization is for! We can create a ParsedJsonData type and a rule that produces it from a Digest, and let the engine do its thing :)

Reply:
ok, I'm convinced :)